parsing/location.ml

```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
+    ~(read_char: unit -> char option):
+  (string (* line *) * int (* beginning of line *)) list
+  =
+  seek start_pos.pos_bol;
+  let lines = ref [] in
+  let bol = ref start_pos.pos_bol in
+  let cur = ref start_pos.pos_bol in
+  let b = Buffer.create 80 in
+  let add_line () =
+    if !bol < !cur then begin
+      let line = Buffer.contents b in
+      Buffer.clear b;
+      lines := (line, !bol) :: !lines;
+      bol := !cur
+    end
+  in
+  let rec loop () =
+    if !bol >= end_pos.pos_cnum then ()
+    else begin
+      match read_char () with
+      | None ->
+          (* end of input *)
+          add_line ()
+      | Some c ->
+          incr cur;
+          match c with
+          | '\r' -> loop ()
+          | '\n' -> add_line (); loop ()
+          | _ -> Buffer.add_char b c; loop ()
+    end
+  in
+  loop ();
+  List.rev !lines
+
+(* Try to get lines from a lexbuf *)
+let lexbuf_lines_around
+    ~(start_pos: position) ~(end_pos: position)
+    (lb: lexbuf):
+  (string * int) list
+  =
+  (* Converts a global position to one that is relative to the lexing buffer *)
+  let rel n = n - lb.lex_abs_pos in
+  if rel start_pos.pos_bol < 0 then begin
+    (* Do nothing if the buffer does not contain the input (because it has been
+       refilled while lexing it) *)
+    []
+  end else begin
+    let pos = ref 0 in (* relative position *)
+    let seek n = pos := rel n in
+    let read_char () =
+      if !pos >= lb.lex_buffer_len then (* end of buffer *) None
+      else
+        let c = Bytes.get lb.lex_buffer !pos in
+        incr pos; Some c
+    in
+    lines_around ~start_pos ~end_pos ~seek ~read_char
+  end
+
+(* Get lines from a file *)
+let file_lines_around
+    ~(start_pos: position) ~(end_pos: position)
+    (filename: string):
+  (string * int) list
+  =
+  try
+    let cin = open_in_bin filename in
+    let read_char () =
+      try Some (input_char cin) with End_of_file -> None
+    in
+    let lines =
+      lines_around ~start_pos ~end_pos ~seek:(seek_in cin) ~read_char
+    in
+    close_in cin;
+    lines
+  with Sys_error _ -> []
+
+(* A [get_lines] function for [highlight_quote] that reads from the current
+   input.
+
+   It first tries to read from [!input_lexbuf], then if that fails (because the
+   lexbuf no longers contains the input we want), it reads from [!input_name]
+   directly *)
+let get_lines_from_current_input ~start_pos ~end_pos =
+  (* Be a bit defensive, and do not try to open one of the possible
+     [!input_name] values that we know do not denote valid filenames. *)
+  let file_valid = function
+    | "//toplevel//" | "_none_" | "" -> false
+    | _ -> true
+  in
+  let get_lines_from_file () =
+    if file_valid !input_name then
+      file_lines_around !input_name ~start_pos ~end_pos
+    else
+      []
+  in
+  match !input_lexbuf with
+  | Some lb ->
+      begin match lexbuf_lines_around lb ~start_pos ~end_pos with
+      | [] -> (* The input is likely not in the lexbuf anymore *)
```

Comment:
eh, why not return a `line list option`?

Reply:
Then `Some []` would also be an error case, to be handled exactly the same way as `[]`. I don't think it's useful to complicate the logic and distinguish between two failure cases that are going to be handled the same.